plugins/primus_lisp/lisp/simple-memory-allocator.lisp

```diff
@@ -29,22 +29,45 @@
 
 (defun memory/allocate (ptr len)
   (if *malloc-initialize-memory*
-      (memory-allocate ptr n *malloc-initial-value*)
-    (memory-allocate ptr n)))
+      (memory-allocate ptr len *malloc-initial-value*)
+    (memory-allocate ptr len)))
+
+(defun word-size () (/ (word-width) 8))
+
+(defun encode-memory-length(ptr len)
+  (write-word ptr_t ptr len))
+
+(defun decode-memory-length(ptr)
+  (read-word ptr_t ptr))
+
+(defun decode-memory-length'(ptr)
+  (if ptr (decode-memory-length (- ptr (word-size))) 0))
 
 (defun malloc (n)
   "allocates a memory region of size N"
   (declare (external "malloc"))
   (if (= n 0) *malloc-zero-sentinel*
     (if (malloc-will-reach-limit n) 0
-      (let ((n (+ n (* 2 *malloc-guard-edges*)))
+      (let ((width (word-size))
+            (full (+ n (* 2 *malloc-guard-edges*) width))
             (ptr brk)
-            (failed (memory/allocate ptr n)))
+            (failed (memory/allocate ptr full)))
         (if failed 0
-          (set brk (+ brk n))
-          (malloc/fill-edges ptr n)
-          (+ ptr *malloc-guard-edges*))))))
-
+          (set brk (+ brk full))
+          (malloc/fill-edges ptr full)
+          (set ptr (+ ptr *malloc-guard-edges*))
+          (encode-memory-length ptr n)
+          (+ ptr width))))))
+
+(defun realloc (ptr new)
```

Comment:
Please, rewrite this function to conform with [POSIX][1]. In particular,
1. when `ptr` is NULL it shall behave as malloc
2. when `len` is NULL it shall behave as free
3. when `len` is unchanged nothing should be done
4. when `len` is smaller the pointer shall not change, but see a not below.
5. only when `len` is greater than the `old-len` we may return a new pointer (and we don't have other choices with our current allocator).
6. if a new chunk is allocated by realloc the old chunk, pointed by `ptr` shall be deallocated.
The following code (untested) implements this semantics. Note, how `(or (and ..) (and ..) ...)` is used in place where you will be using pattern matching in OCaml. This is much easier to read than a bunch of nested if/then/elses. 


```lisp
(defun realloc (ptr len)
  (or (and (not ptr) (malloc ptr len))
      (and (not len) (free ptr))
      (let ((old-len (malloc/get-chunk-size ptr)))
        (if (<= old-len len)
            (prog (malloc/put-chunk-size ptr len) ptr)
          (free ptr)
          (let (new-ptr (malloc ptr len))
            (if new-ptr (memcpy new-ptr ptr old-len)
              new-ptr))))))
```

Note, if `realloc` is used to reduce the size of a chunk (e.g., for shrinking a dynamic vector), we shall return the same pointer, but mark the region `ptr+len,ptr+old-len` as invalid. Therefore, we need to update the region-analysis (where malloc is hooked into it, and track calls to realloc which shrink the regions and try to mark them as freed). If it would be too hard, we can ignore it right now and think about it later.

[1]:  https://pubs.opengroup.org/onlinepubs/009695399/functions/realloc.html

Reply:
This `realloc` is also buggy, by the way, the trick `(or (and <cond> <body>) ...)` doesn't work here, since `<body>` in our case could also evaluate to false so the control flow will fall down, i.e., both `malloc` and `free` could return false, in fact `free` always returns false, so we will fall after free to realloc. It won't break a lot, i.e., we will now hit the `(<= old-len len)` guard, and just write `0` to the length field. But still it is more a bug than a feature.